servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpRequesterFilter.java

```diff
@@ -100,4 +167,59 @@ public HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strat
         // No influence since we do not block.
         return strategy;
     }
+
+    @FunctionalInterface
+    public interface TimeoutFromRequest extends Function<HttpRequestMetaData, Duration> {
+
+        /**
+         * Determine timeout duration, if present, from a request and/or apply default timeout durations.
+         *
+         * @param request the current request
+         * @return The timeout or null for no timeout
+         */
+        @Override
+        @Nullable
+        Duration apply(HttpRequestMetaData request);
+    }
+
+    /**
+     * Returns a function which uses {@link #useContextDeadlineOrDefault(Duration) useContextDeadlineOrDefault()} with
+     * the provided default duration to determine the timeout duration to be used for a provided request.
+     *
+     * @param duration default timeout duration or null for no timeout
+     * @return a function to produce a timeout based on the context deadline or specified default
+     */
+    public static TimeoutFromRequest useDefaultTimeout(@Nullable Duration duration) {
+        // the request is not considered
+        return (req) -> useContextDeadlineOrDefault(duration);
+    }
+
+    /**
+     * Returns timeout duration calculated based on the remaining time until the deadline in the context or, if absent,
+     * the provided default. The timeout, if any, will be added to the context as a deadline for additional client
+     * requests initiated within this context. The contents of the request are not used by this implementation.
+     *
+     * <p><strong>Note:</strong>This implementation assumes that the {@link AsyncContext} has not been
+     * {@link AsyncContext#disable() disabled}, but will always use the default duration if it has been disabled.
+     *
+     * @param defaultDuration default timeout duration or null for no timeout
+     * @return a timeout based on the context deadline or specified default (which may be null)
+     */
+    public static @Nullable Duration useContextDeadlineOrDefault(@Nullable Duration defaultDuration) {
+        Instant deadline = AsyncContext.get(HTTP_DEADLINE_KEY);
+
+        if (null != deadline) {
+            return Duration.between(Instant.now(), deadline);
+        } else if (null != defaultDuration) {
+            // Convert it to a deadline so that if reused for a new client request the timeout will reflect actual
+            // remaining time.
+            try {
+                AsyncContext.put(HTTP_DEADLINE_KEY, Instant.now().plus(defaultDuration));
+            } catch (UnsupportedOperationException ignored) {
+                // ignored
```

Comment:
any reason we don't re-throw this rather than silently swallow? Also in the future when we have RequestContext we don't really want folks to use this method IIUC. perhaps we should just internalize this in gRPC where required for now.

Reply:
+1 for moving this logic to gRPC. At HTTP level, an ability to provide a function is enough. If users need deadlines for HTTP, they can implement their function to support it. There is no much complexity in this function and it probably won't fit all use-cases.

Reply:
In prior discussions the important part was to ensure that client requests made by servers were bound by the active time limit. In basic HTTP this will come from the TimeoutServiceFilter but in gRPC it will propagate from the originating client to the server and then be incorporated in to client requests.

Keep this feature for HTTP or not worth it? The issue for servers making client requests is determining the amount of time still available and the async context still seems to be the best option for that.

Reply:
Regarding not throwing: I discovered needing to catch the UnsupportedOperationException when I disabled the AsyncContext. The problem with re-throwing is that doing so would break the usage/application completely. Breaking the feature, passing the timeout in the async context, seemed preferable and a reasonable response to the AsyncContext being disabled.